### ISSUE TITLE
Applying rotation to HuggingFace model

### ISSUE THREAD

**Issue by YLGH** [spcl/QuaRot#1]
First off, thanks for the great paper that makes these concepts super approachable!

I was playing around with the code base and rotate some models to get some intuition about what's happening

I'm running

import os
import tempfile
import torch
import transformers

```
device = torch.device("cuda")
model_path = "/data/hf-models/Llama-2-7b-hf/"
tokenizer = transformers.AutoTokenizer.from_pretrained(model_path)
model = transformers.AutoModelForCausalLM.from_pretrained(model_path, torch_dtype=torch.float16).to(device)
fuse_layer_norms(model)
quarot.rotate_model(model)
```

However, afterward the model outputs don't match up with the pre-rorated model (it looks like gibberish). Is this something that's expected to work?

Also, should fuse_ln_linear be used for a Llama model which uses RMS norm? 

Thanks!

**Comment by YLGH** [comment]
Ah I figured out my issue, had a bug where I didn't rotate after attention,

I'm playing with a model that does step 1 of the paper e.g.:
1. merge weights of RMSNorm into linear weights, so it's effectively only x <- ||x||
2. Apply one side Q/Q.T matrix

I can tell that the incoherence of the new model is much better and seems to produce english. However, I ran MMLU on the the original llama-2-7b-hf model and got 

Perplexity  1.41031
Accuracy    0.482624

But running on the new rotated matrix and got

Perplexity  1.65062
Accuracy    0.430637

, the .05 accuracy drop seems pretty huge. Any ideas what could've gone wrong? My first thought is that since this is a 7b model maybe it is harder to quantize...or maybe the eps in the RMSNorm makes a big difference (I guess in theory it's not 100% mathematically equivalent with that there)

**Comment by sashkboos** [comment]
Hi @YLGH 

Thank you so much for using our code. 

This maybe because of numerical issues. However, can you please write what you have tried (all steps) and provide the code here so I can reproduce the issue (and the numbers)?

Thanks

**Comment by YLGH** [comment]
@sashkboos  

Here's my full repro for exporting the model:

https://gist.github.com/YLGH/a520ce4214545bc06c97e9b5ea1d7fc9

My check_model() response is pretty much identical to the original, just when I run a MMLU eval against it does it seem to go off

**Comment by YLGH** [comment]
My guess is that there's something weird with numerics that happen when I merge weights from rms norm into linear, 

**### Original model**
is when it is ripe.
The best time to eat a strawerry is when it is ripe. The best time to eat a strawerry is when it is ripe. The best time to eat a strawerry is
**### After merging rms with linears**
is when it is freshly picked.

### 📚

- [The best time to eat a strawerry is when it is freshly picked](https://www.reddit.com/r/programming
**### After rotating**
is when it is freshly picked.

### 📚

- [The best time to eat a strawerry is when it is freshly picked](https://www.reddit.com/r/programming

**Comment by YLGH** [comment]
Ah, this was all my fault, sorry for the churn! I don't need to move the .mean() to the linear weights,

**Comment by sashkboos** [comment]
Thanks @YLGH for using our code. Glad to hear that you fixed the issue. I will close this issue

**Comment by zhsky2017** [comment]
> @sashkboos
> 
> Here's my full repro for exporting the model:
> 
> https://gist.github.com/YLGH/a520ce4214545bc06c97e9b5ea1d7fc9
> 
> My check_model() response is pretty much identical to the original, just when I run a MMLU eval against it does it seem to go off.
> 
> Thanks in advance!

Hi YLGH，Can you provide the corrected code? I ran the code in this link and printed the model output and found that the rotated model was not aligned with the original model (not the generated token id, but the original output tensor of the model)

**Comment by zhsky2017** [comment]
> @sashkboos
> 
> Here's my full repro for exporting the model:
> 
> https://gist.github.com/YLGH/a520ce4214545bc06c97e9b5ea1d7fc9
> 
> My check_model() response is pretty much identical to the original, just when I run a MMLU eval against it does it seem to go off.
> 
> Thanks in advance!

I noticed that the code you provided is slightly different from the implementation in the original warehouse. For example, you commented out some transformations in the original warehouse. Can you explain why? (I tried to run fake_quant/main.py to get the rotated model, but I tested ppl and found that the rotated model results were very poor. The ppl obtained using your code is very close to the original model)

**Comment by sashkboos** [comment]
@zhsky2017 

We used the original implementation from GPTQ [repo](https://github.com/IST-DASLab/gptq) (which is widely used in the literature) for generating our baseline model (and evaluating the PPL). We only added the batch dimension for PPL calculation to the original code. You can check our evaluation function which is implemented [here](https://github.com/spcl/QuaRot/blob/main/fake_quant/eval_utils.py#L11)

**Comment by zhsky2017** [comment]
> @zhsky2017
> 
> We used the original implementation from GPTQ [repo](https://github.com/IST-DASLab/gptq) (which is widely used in the literature) for generating our baseline model (and evaluating the PPL). We only added the batch dimension for PPL calculation to the original code. You can check our evaluation function which is implemented [here](https://github.com/spcl/QuaRot/blob/main/fake_quant/eval_utils.py#L11)

Sorry, I may not have described my problem clearly. What I want is the model after rotation (not quantized). According to the description in your paper, this transformation should be identical. I want to verify the model before and after the transformation (not quantized). Identity, but when I run fake_qunt/main.py (removing the quantization part and only retaining the transformation) to obtain the rotated model and compare it with the original model, their outputs are inconsistent.

**Comment by sashkboos** [comment]
@zhsky2017 

Are you using the following command? 

```
/bin/python main.py --model meta-llama/Llama-2-7b-hf  --rotate
```
This results the same PPL for me in the code. I would suggest to use the same versions of the packages in the repo.

If this is not the case, please provide the exact setting so I can reproduce the issue and take a look at it. 

Thanks

**Comment by YLGH** [comment]
@zhsky2017 

Here's what I have going so far: https://gist.github.com/YLGH/2563ef7387633e71477865755fb34006

@sashkboos After I used the fuse_ln_linear from the Github repo, I did notice that the model outputs had some poor results. I figured it might've been some complexity with LayerNorm?
